src/lib.rs

```diff
@@ -0,0 +1,924 @@
+            index += indices[k] * stride;
+        }
+        index
+    }
+
+    fn assert_indices(&self, indices: &[usize]) -> Result<(), String> {
+        if indices.len() != self.shape.len() {
+            return Err(format!("Incorrect number of dimensions ({} vs {}).", indices.len(), self.shape.len()));
+        }
+        for (i, &index) in indices.iter().enumerate() {
+            if index >= self.shape.dims[i] {
+                return Err(format!("Index out of bounds for dimension {}", i));
+            }
+        }
+        Ok(())
+    }
+
+}
+
+// Element-wise Multiplication
+impl<T: Num + PartialOrd + Copy> Mul<T> for Tensor<T> {
+    type Output = Tensor<T>;
+
+    fn mul(self, rhs: T) -> Tensor<T> {
+        let mut result = Tensor::zeros(&self.shape);
+        for i in 0..self.size() {
+            result.data[i] = self.data[i].clone() * rhs;
+        }
+        result
+    }
+}
+
+// Vector/Matrix Multiplication
+impl<T: Num + PartialOrd + Copy> Mul<Tensor<T>> for Tensor<T> {
+    type Output = Tensor<T>;
+
+    fn mul(self, rhs: Tensor<T>) -> Tensor<T> {
+    if self.shape.len() == 1 && rhs.shape.len() == 1 {
+        // Vector-Vector multiplication (dot product)
+        assert!(self.shape[0] == rhs.shape[0], "Vectors must be of the same length for dot product.");
+        let mut result = T::zero();
+        for i in 0..self.shape[0] {
+            result = result + self.data[i] * rhs.data[i];
+        }
+        Tensor::new(&shape![1], &vec![result])
+    } else if self.shape.len() == 1 && rhs.shape.len() == 2 {
+        // Vector-Matrix multiplication
+        assert!(self.shape[0] == rhs.shape[0], "The length of the vector must be equal to the number of rows in the matrix.");
+        let mut result = Tensor::zeros(&shape![rhs.shape[1]]);
+        for j in 0..rhs.shape[1] {
+            let mut sum = T::zero();
+            for i in 0..self.shape[0] {
+                sum = sum + self.data[i] * rhs.data[i * rhs.shape[1] + j];
+            }
+            result.data[j] = sum;
+        }
+        result
+    } else if self.shape.len() == 2 && rhs.shape.len() == 1 {
+        // Matrix-Vector multiplication
+        assert!(self.shape[1] == rhs.shape[0], "The number of columns in the matrix must be equal to the length of the vector.");
+        let mut result = Tensor::zeros(&shape![self.shape[0]]);
+        for i in 0..self.shape[0] {
+            let mut sum = T::zero();
+            for j in 0..self.shape[1] {
+                sum = sum + self.data[i * self.shape[1] + j] * rhs.data[j];
+            }
+            result.data[i] = sum;
+        }
+        result
+    } else if self.shape.len() == 2 && rhs.shape.len() == 2 {
+        // Matrix-Matrix multiplication
+        assert!(self.shape[1] == rhs.shape[0], "The number of columns in the first matrix must be equal to the number of rows in the second matrix.");
+        let mut result = Tensor::zeros(&shape![self.shape[0], rhs.shape[1]]);
```

Comment:
Broadcasting and Tensor slices would be next on the to do list.

Reply:
Speaking of a todo list...

We don't have one.

It will be very easy to get carried away and just implement a bunch of stuff. We should come up with a set of operations to stop at and add more later if there's appetite.

I will start one in an issue.

Reply:
Yeah I guess that's what we can use this initial implementation for.

Reply:
https://github.com/Rust-Scientific-Computing/feotensor/issues/1#issuecomment-2183283636